src/lib.rs

```diff
@@ -307,23 +307,43 @@ pub trait Resampler<T>: Send {
     /// the buffer size is only guaranteed to be sufficient to prevent allocation
     /// within [Resampler::process_into_buffer] until the resampling ratio is changed.
     fn allocate_output_buffer(&self) -> Vec<Vec<T>> {
-        let (channels, out_len) = self.get_max_output_size();
-        let mut wave_out = Vec::with_capacity(channels);
+        let frames = self.get_max_output_frames();
+        let channels = self.nbr_channels();
+        let mut buffer = Vec::with_capacity(channels);
         for _ in 0..channels {
-            wave_out.push(Vec::with_capacity(out_len));
+            buffer.push(Vec::with_capacity(frames));
         }
-        wave_out
+        buffer
     }
 
-    /// Get the max output size the resampler can give as (channels, frames).
+    /// Get the max number of output frames per channel.
     /// For the Resamplers with variable output sizes ([FftFixedIn] and [SincFixedIn]),
     /// the max output size is only guaranteed to be sufficient to prevent allocation
     /// within [Resampler::process_into_buffer] until the resampling ratio is changed.
-    fn get_max_output_size(&self) -> (usize, usize);
+    fn get_max_output_frames(&self) -> usize;
+
+    /// Get the maximum number of channels this Resampler can process
+    fn nbr_channels(&self) -> usize;
 
     /// Query for the number of frames needed for the next call to "process".
     fn nbr_frames_needed(&self) -> usize;
 
+    /// Get the maximum number of input frames the resampler could require
+    fn max_nbr_frames_needed(&self) -> usize;
+
+    /// Convenience method for allocating an input buffer large enough to be passed to
```

Comment:
> This description isn't quite right. The function creates a buffer with sufficient capacity to avoid reallocating. Then the length needs to be set correctly for each call to process. The WrongNumberOfInputSamples means that the length, not the capacity, of the input is wrong.

Good point. I fixed the documentation.

> Another point here, what capacity should it allocate? In practice this only matters for SincFixedOut. There it currently allocates enough space for the full adjust range. For applications that use adjust to rate-match, this basically means it's allocating 10 times the space needed. Seems a bit wasteful! It's also different from allocate_output_buffer which only reserves space for the current ratio. I think these two should be consistent.

Good point about the inconsistency between `allocate_input_buffer` and `allocate_output_buffer`. I think this should be resolved by making `allocate_output_buffer` allocate the maximum capacity possibly needed instead of making it dependent on the current ratio, which I have implemented.

As for wasting memory, for any realtime application we're talking about a few kilobytes of RAM. Wasting a few kilobytes RAM is no big deal compared to heap allocating in a realtime thread. Maybe for some other non-realtime application the pros and cons of this tradeoff are not the same. I think the best way to resolve this is to let the user specify the maximum ratio range upon construction, which I have implemented.

Reply:
Yes is really depends on the application if it's better to waste ram or do a few allocations. And that's a good argument for making it configurable, making it worth providing the extra parameter when creating the resampler.